### ISSUE TITLE
push! does not seem to be thread safe

### ISSUE THREAD

**Issue by alessandrousseglioviretta** [dandeliondeathray/DandelionWebSockets.jl#12]
After having failed to send text by using `send_text` from within a `on_binary` function (no text is sent), I dug a little into the code and found out that in _utilities.jl_, `push!(m.actuals, (s, collect(args)))` is probably not thread safe, which could be the cause of the failure.

The problem needs to be confirmed, a good starting point being the following discussion https://discourse.julialang.org/t/thread-safe-array-building/3275

Thanks!

**Comment by erikedin** [comment]
I made a fix just now, which could possibly/probably be the issue you had. The commit is 7a49b94 "copy payload before masking it".
The problem was that the payload in a WebSocket message was not copied, so the original data you sent was modified (as each message is XOR'd with a key). This will definitely lead to invalid UTF-8 data being sent sometimes, which the WebSocket server could handle by ignoring the message. Actually, the server should close the connection when received such an invalid message, but at least one server I tried did not.
The utilities.jl file you mentioned is part of the tests, so it should not impact the production code, although it may well be an issue for the tests.

I'll leave this issue open for the time being, until you're satisfied it's fixed.
